### PR TITLE
Sync SisImport refactoring - automated tests for it

### PR DESCRIPTION
**Pull request by sadleb**
# Description
We released the new sync architecture here:
* https://github.com/bebraven/platform/pull/993

But we didn't finish all the automated tests. This PR adds the missing ones.

# Motivation and Context
Task: https://app.asana.com/0/1201131148207877/1200769578679244

# Code Review notes
For the actual `spec` files, the diff isn't helpful since I just had scratch notes in there.
I would just choose "View File" and review the file directly:
image

Files to review in approximate order (skip the rest)
* app/services/sync_salesforce_program.rb
* app/models/heroku_connect/program.rb
* spec/services/sync_salesforce_program_spec.rb
* app/services/sync_salesforce_participant.rb
* spec/services/sync_salesforce_participant_spec.rb
* spec/services/sync_salesforce_contact_spec.rb
* app/models/participant_sync_info.rb
* spec/models/participant_sync_info_spec.rb

### CODE REVIEW

File: app/services/sync_salesforce_participant.rb [hunk 1]
```diff
@@ -95,22 +95,23 @@ def set_accelerator_course_enrollments
   end
 
   def set_lc_playbook_course_enrollments
-    return unless @participant_sync_info.is_enrolled?
-
```
Comment (sadleb) [inline app/services/sync_salesforce_participant.rb:99]: Bug fix. Wasn't dropping local LC Playbook section roles.

File: app/services/sync_salesforce_program.rb [hunk 2]
```diff
@@ -106,13 +106,13 @@ def run
...
-      # merge. If no other Participants had changes that actually got through the sync,
-      # we'd skip sending the same thing to Canvas that we just sent.
-      if @synced_new_canvas_changes
+      # Only send the SIS Import to Canvas if we force it or there are new changes that
+      # actually synced. This is to prevent continuously sending the same data to Canvas
+      # every sync if we have failing Participants that take some time for us to fix the
+      # underlying issue. For example, imagine we had one with a duplicate Contact and it
+      # took us a day to merge. If no other Participants had changes that actually got
+      # through the sync, we'd skip sending the same thing to Canvas that we just sent.
+      if @force_canvas_update || @synced_new_canvas_changes
```
Comment (sadleb) [inline app/services/sync_salesforce_program.rb:115]: Bug fix. Wasn't sending to Canvas in `@force_canvas_update` unless there were actual changes.

**Review by elenabaurkot — COMMENTED** [review]
Lgtm 👍

File: spec/services/sync_salesforce_participant_spec.rb [hunk 3]
```diff
@@ -1,471 +1,600 @@
...
+    cohort_section: cohort_section
+  }
+
+  before(:each) do
+    allow(SyncSalesforceContact).to receive(:new).and_return(sync_contact_service)
+    allow(sync_contact_service).to receive(:run).and_return(user)
+    allow(CreateSection).to receive(:new).and_return(create_section_service)
+    allow(create_section_service).to receive(:run)
+    allow(Honeycomb).to receive(:add_alert)
+    allow(Honeycomb).to receive(:add_field)
```
Comment (rshipp) [inline spec/services/sync_salesforce_participant_spec.rb:45]: are these used for something? i don't see them referenced below

**Review by rshipp — APPROVED** [review]

File: spec/services/sync_salesforce_contact_spec.rb [hunk 4]
```diff
@@ -3,214 +3,326 @@
...
     end
 
-    context 'when SignupDate__c fails to send to Salesforce' do
-      xit 'does not raise error' do
-        # TODO: this won't cause end-user issues, so don't have it fail for them
+    shared_examples 'when Platform email matches Salesforce' do
+      shared_examples 'is a NOOP' do
+        it 'leaves the user unchanged' do
+          run_sync
+          expect(User.find(user.id)).to eq(@user_before_sync)
```
[the remark this answers is not in the document]
Comment (sadleb) [reply spec/services/sync_salesforce_contact_spec.rb]: artifact of previous attempts. removed.

File: spec/services/sync_salesforce_participant_spec.rb (hunk 3 above)
Comment (sadleb) [reply spec/services/sync_salesforce_participant_spec.rb]: removed

**Pull request merged**